### DISCUSSION TITLE
Mocking

### DISCUSSION THREAD

**Discussion by atakankarsli** [MobileNativeFoundation/discussions#158] [General]
Since most of the current mobile testing solutions too much depend on mocks, I want to hear from folks about your mocking strategies.  How/Where do you add, serve and maintain them? 

Trendyol currently uses :
1-) Swifter as a mock server to supply responses for Smoke/UI Tests. [Mocking API Calls in UI Tests](https://medium.com/trendyol-tech/mocking-api-calls-in-ui-tests-b908dbf6b305)
2-) We have separated mock JSON files for both Unit and UI tests. Unit mocks are located under the same module with unit tests. And UI test JSONs are located inside the UITest folder and we maintain them separately.
3-) For adding new mocks we have MockServerHandler which saves all responses inside the simulator so we can drag them inside the project and add a necessary mapping.

We are planning to combine them to use the same mocks with Unit tests since it will be easier to maintain and be sure about coverage. Not sure where we should put them yet. 

From previous discussions, I'm curious about your strategies specifically thanks in advance for your answers. 
@keith  @dflems @fdiaz @noahsmartin

**Comment by keith** [comment]
We have a few pieces in our setup for this:

1. Our networking stack is generated via protobuf, and as part of this generation, we generate a mocks module for each different endpoint and the types that go with them. Developers can pull these in to their tests and use them. They are entirely separate modules so they do not affect code coverage or anything as we do not care about how much of the generated code is covered. This is super useful because the mocks are always just there for devs when they need them.
2. We have another module type for business logic specific mocks. This module lives alongside the feature / tests, but also doesn't affect code coverage. These modules can be included by the specific feature's tests, as well as from tests up the dependency tree in other features. This provides a nice way for developers to provide mocks for their APIs that can be consumed in many different places (which wouldn't work the same way if they were part of their test modules).
3. This mocks module also allows for generating mocks from protocols in the source module. We use something like sourcery for this. While theoretically this could entirely replace the need for handwritten mocks in these modules, there's some nice use cases where developers provide defaulted initializers and similar helpers that make testing easier, even though those are a bit less like mocks and more like module specific test helpers.
4. We have some internal infrastructure around mocking real server responses and things at that level, that we use for UI tests, this involves spinning up some fake response handler with the expected response and running the test against the fake server instead of an actual one.
